### DISCUSSION TITLE
Running fits on HTCondor

### DISCUSSION THREAD

**Discussion by ssaha1234** [scikit-hep/pyhf#1807] [Q&A] [answered]
Dear pyhf users,
   I would like to submit pyhf jobs on HTCondor for doing mle fits. I can run the fits locally by doing the following steps - 

```console
$ source setup.sh
$ python Fit.py
```

* `setup.sh`: https://gist.github.com/ssaha1234/d504ce501bbc1e126f47dbb7233c51b2
* `req.txt`: https://gist.github.com/ssaha1234/eace318069d44f0ed9de041ed937b1d9
* `Fit.py`: https://gist.github.com/ssaha1234/0e84c80cd22fb36126c7302462170ae6
* Condor submission script: https://gist.github.com/ssaha1234/36823f9620b11e91c9292d2b0b15580b

The error file, makes sense to me, since I am not setting up `pyhf` on the machine, the job is submitted to - 

```pytb
Traceback (most recent call last):
File "condor_exec.exe", line 2, in <module>
   import pyhf
ModuleNotFoundError: No module named 'pyhf'
```

Is there a way to setup pyhf on condor and submit the jobs ? I am thinking if we have to write a shell script wrapper for it ?
Thank you !
Best, 
Shreya

**Comment by matthewfeickert** [comment, marked as answer]
> Is there a way to setup pyhf on condor and submit the jobs ? I am thinking if we have to write a shell script wrapper for it?

There's nothing specific to `pyhf` here. If you want to load a particular environment then you need to ensure that environment gets loaded in your HTCondor submit file's `Executable`. So if you make your `Executable` a shell script (e.g. `example.sh`) you would already have generated your Python virtual environment in advance in a location that is accessible by HTCondor (so you're not creating it over and over each time you run) and then source that Python virtual environment before you do anything.

For example, if you were on a CVMFS machine and had a Python virtual environment built from a `requirements.txt` and `create_cvmfs_venv.sh`

<details>
<summary>create_cvmfs_venv.sh:</summary>

```bash
#!/bin/bash

# load LCG view to get Python
. /cvmfs/sft.cern.ch/lcg/views/LCG_101/x86_64-centos7-gcc11-opt/setup.sh
# Create Python virtual environment
python -m venv venv
. venv/bin/activate
# Ensure that the Python virtual environment is at head of PYTHONPATH
# c.f. https://github.com/matthewfeickert/cvmfs-venv for more info and alternatives to deal with this
export PYTHONPATH="$(readlink -f $(find venv -type d -iname site-packages)):${PYTHONPATH}"

python -m pip install --upgrade --ignore-installed pip setuptools wheel
python -m pip install --upgrade --ignore-installed pip-tools

# Generate environment lock file
pip-compile \
    --generate-hashes \
    --output-file requirements.lock \
    requirements.txt

# Use Brett Cannon's recommendations for pip-secure-install to ensure environment
# is reproducible and installed as securely as possible.
# c.f. https://www.python.org/dev/peps/pep-0665/#secure-by-design
# c.f. https://github.com/brettcannon/pip-secure-install
# c.f. https://twitter.com/brettsky/status/1486137764315688961
# As working in CVMFS LCG views also add --ignore-installed to avoid CVMFS packages
python -m pip install \
    --ignore-installed \
    --no-deps \
    --require-hashes \
    --only-binary :all: \
    --requirement requirements.lock

python -m pip list --local
```

</details>

then if you had your executable be something like `example.sh` which calls `example.py`

<details>
<summary>example.sh:</summary>

```bash
#!/bin/bash

# Load environment
. /cvmfs/sft.cern.ch/lcg/views/LCG_101/x86_64-centos7-gcc11-opt/setup.sh  # load LCG view to get Python
. /afs/cern.ch/user/f/feickert/htcondor_example/venv/bin/activate  # load Python virtual environment
# Ensure that the Python virtual environment is at head of PYTHONPATH
# c.f. https://github.com/matthewfeickert/cvmfs-venv for more info and alternatives to deal with this
# This should be the equivalent of: export PYTHONPATH="$(readlink -f $(find venv -type d -iname site-packages)):${PYTHONPATH}"
export PYTHONPATH="/afs/cern.ch/user/f/feickert/htcondor_example/venv/lib/python3.9/site-packages:${PYTHONPATH}"

# Run script
python example.py
```

</details>

you should be able to submit it in your HTCondor submission script and have things run.

Here's a [GitHub Gist with a full example](https://gist.github.com/matthewfeickert/3c33297e087fb0c066301cd06e31c52f).

(Note that I wrote this without running it and I'm more of a SLURM user, so I might have made a mistake.)

**Reply by ssaha1234** [reply]
Dear @matthewfeickert, 
   Thank you very much for the examples, yes this is indeed not specific to pyhf, it is about loading the right environment. I will look through the examples and work it out :))
Best, 
Shreya

**Reply by matthewfeickert** [reply]
@kratsg has pointed out to me that it isn't safe to assume that HTCondor will have any access to the user's file system at runtime and that the better way to do this is to have a shell script that tars up the existing Python virtual environment and then provides them as `transfer_input_files` to the submit file.

**Reply by ssaha1234** [reply]
Dear @matthewfeickert, 
   Yes, indeed. I just got the error, that example.py cannot be found bu condor,  and I need to transfer input files ! 
I am going to rerun and check. 
Thanks @kratsg for the hack ! 
Best, 
Shreya

**Reply by ssaha1234** [reply]
Hi @matthewfeickert, 
   It works now ! Thanks a lot for the GitHub gist. 
Best, 
Shreya

**Reply by matthewfeickert** [reply]
> I just got the error, that example.py cannot be found bu condor, and I need to transfer input files

Yeah, I had forgot to account for that last night while writing it. I've corrected it now so the [GitHub Gist](https://gist.github.com/matthewfeickert/3c33297e087fb0c066301cd06e31c52f) should be updated.

> It works now ! Thanks a lot for the GitHub gist

Glad to hear it, but @kratsg's point was different than just adding

```
transfer_input_files  = example.py
```

Here things are working on CERN's instance of HTCondor deployed on LXPLUS, but that's more by luck than design. There are no general guarantees that HTCondor will be able to have read access to a user's disk space while running jobs, so you should find a way to transfer the **entire runtime environment** instead.

**Reply by ssaha1234** [reply]
Dear @matthewfeickert and @kratsg, 
       Thanks for creating the github gist with the example. I am running a simple CLs code on Condor, and I have followed the code in the gist. However, when I print out the version of pyhf in my code, it shows 0.5.4, and it is supposed to be 0.6.3, according to the `requirements.txt` file. Also, I have added the following lines in the `requirements.txt` file for using `jax`.

`pyhf==0.6.3`
`mplhep`
`numpy`
`matplotlib`
`jax==0.3.16`
`jaxlib==0.3.15
`

But for the `jax` part, it is showing an error - 
`ERROR: Could not find a version that satisfies the requirement jax==0.3.16 (from versions: 0.1.19, 0.1.20, 0.1.21, 0.1.22, 0.1.23)`
`ERROR: No matching distribution found for jax==0.3.16`. Is there something I am missing here ?

Thanks ! 
Best, 
Shreya

**Reply by kratsg** [reply]
Your python version is probably 3.6 (we only support 3.7+). You need to use a non-lxplus-default python.

**Reply by ssaha1234** [reply]
Hi @kratsg, 
    Thanks for your reply ! I see that my python version is 3.8.6 (when I type `python --version ` ). I still see the same jax error, after running the `create_cvmfs_venv.sh` file. I replaced the following line - 
`# load LCG view to get Python`
`. /cvmfs/sft.cern.ch/lcg/views/LCG_101/x86_64-centos7-gcc11-opt/setup.sh`

to 

`export ATLAS_LOCAL_ROOT_BASE=/cvmfs/atlas.cern.ch/repo/ATLASLocalRootBase`
`export ALRB_localConfigDir=$HOME/localConfig`
`source ${ATLAS_LOCAL_ROOT_BASE}/user/atlasLocalSetup.sh`
`lsetup "python 3.7.4-x86_64-centos7"`

Maybe I am missing something more here ?
Thanks ! 
Best, 
Shreya

**Reply by kratsg** [reply]
Using [cvmfs-venv](https://github.com/matthewfeickert/cvmfs-venv) from @matthewfeickert, this works for me just fine:

```
curl -sLO https://raw.githubusercontent.com/matthewfeickert/cvmfs-venv/main/atlas_setup.sh
. atlas_setup.sh testssha1234
python -m pip install -r requirements.txt
```

The issue is just with how lxplus deals with python virtual environments (which is not very well)

**Reply by ssaha1234** [reply]
Hi @kratsg ,  
    Thank you very much for the check, I am not using lxplus in this case, I should have mentioned it before, when I asked the question. 
I am using the ATLAS analysis facility at UChicago. Hence, I think there might be some new changes in above example ?
Best, 
Shreya

**Reply by kratsg** [reply]
This also works on any machine with CVMFS. So try it out.

**Reply by ssaha1234** [reply]
Hi @kratsg, 
   Thanks for letting me know, I will try it out from scratch, from the github gist :)
Best, 
Shreya

**Reply by ssaha1234** [reply]
Hi @kratsg, 
   It all worked for me now ! Thank you very much ! 
Best, 
Shreya
